framework/src/kernels/ADArrayReaction.C

```diff
@@ -0,0 +1,55 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADArrayReaction.h"
+
+registerMooseObject("MooseApp", ADArrayReaction);
+
+InputParameters
+ADArrayReaction::validParams()
+{
+  InputParameters params = ADArrayKernel::validParams();
+  params += ADFunctorInterface::validParams();
+  params.addRequiredParam<MaterialPropertyName>(
+      "reaction_coefficient",
+      "The name of the reaction, can be scalar, vector, or matrix material property.");
+  params.addClassDescription(
+      "An automatic differentiation kernel for the array reaction operator ($u$)");
+  return params;
+}
+
+ADArrayReaction::ADArrayReaction(const InputParameters & parameters)
+  : ADArrayKernel(parameters),
+    _r(hasMaterialProperty<ADReal>("reaction_coefficient")
+           ? &getMaterialProperty<Real>("reaction_coefficient")
```

Comment:
```suggestion
           ? &getMaterialProperty<ADReal>("reaction_coefficient")
```
should be the same for all types imo